File: atrs-python/src/atrs/api/v1/flights.py
"""Flights API router.

Ported from FlightRestController.java
"""

from datetime import date

from fastapi import APIRouter, HTTPException, Query, status
from pydantic import BaseModel

from ...api.deps import DatabaseDep
from ...core.exceptions import FlightNotFoundException, InvalidSearchCriteriaException
from ...models.enums import BoardingClassCd, FareTypeCd, FlightType
from ...services.b0 import TicketSearchService
from ...services.b0.ticket_search_service import (
    FareTypeVacantInfo,
    FlightVacantInfo,
    TicketSearchCriteria,
)

router = APIRouter()


class FareTypeResource(BaseModel):
    """Fare type with vacancy info."""
    fare_type_cd: FareTypeCd
    fare_type_name: str
    fare: int
    vacant_num: int


class FlightResource(BaseModel):
    """Flight with vacancy info."""
    departure_date: date
    flight_name: str
    departure_time: str
    arrival_time: str
    departure_airport_name: str
    arrival_airport_name: str
    boarding_class_cd: BoardingClassCd
    fare_types: list[FareTypeResource]


@router.get("", response_model=list[FlightResource])
async def search_flights(
    departure_date: date = Query(..., description="Departure date"),
    dep_airport_cd: str = Query(..., min_length=3, max_length=3, description="Departure airport code"),
    arr_airport_cd: str = Query(..., min_length=3, max_length=3, description="Arrival airport code"),
    boarding_class_cd: BoardingClassCd = Query(..., description="Boarding class"),
    flight_type: FlightType = Query(FlightType.OW, description="Flight type"),
    db: DatabaseDep = None,
):
    """Search available flights.
    
    Args:
        departure_date: Flight date
        dep_airport_cd: Departure airport code (e.g., 'HND')
        arr_airport_cd: Arrival airport code (e.g., 'ITM')
        boarding_class_cd: Boarding class (N=Normal, S=Special)
        flight_type: One-way or round-trip
        db: Database connection
        
    Returns:
        List of available flights with fare options
        
    Raises:
        HTTPException 400: If search criteria is invalid
        HTTPException 404: If no flights found
    """
    try:
        criteria = TicketSearchCriteria(
            departure_date=departure_date,
            dep_airport_cd=dep_airport_cd,
            arr_airport_cd=arr_airport_cd,
            boarding_class_cd=boarding_class_cd,
            flight_type=flight_type,
        )
        
        service = TicketSearchService(db)
        results = await service.search_flight(criteria)
        
        return [_to_flight_resource(f) for f in results]
        
    except InvalidSearchCriteriaException as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=e.message,
        )
    except FlightNotFoundException:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No flights found matching criteria",
        )


def _to_flight_resource(flight: FlightVacantInfo) -> FlightResource:
    """Convert FlightVacantInfo to API resource."""
    return FlightResource(
        departure_date=flight.departure_date,
        flight_name=flight.flight_name,
        departure_time=flight.departure_time,
        arrival_time=flight.arrival_time,
        departure_airport_name=flight.departure_airport_name,
        arrival_airport_name=flight.arrival_airport_name,
        boarding_class_cd=flight.boarding_class_cd,
        fare_types=[
            FareTypeResource(
                fare_type_cd=ft.fare_type_cd,
                fare_type_name=ft.fare_type_name,
                fare=ft.fare,
                vacant_num=ft.vacant_num,
            )
            for ft in flight.fare_type_list
        ],
    )

File: JavaConfig-JSP/atrs/atrs-mcp-server/src/tools/flights.ts
/**
 * Flight search tool for ATRS MCP server.
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { makeApiRequest, handleApiError } from "../api-client.js";
import { AIRPORT_CODES, CHARACTER_LIMIT } from "../constants.js";

/** Zod schema matching the ATRS FlightSearchQuery parameters. */
const SearchFlightsSchema = z
  .object({
    flightType: z
      .enum(["OW", "RT"])
      .describe("Flight type: OW (one-way) or RT (round-trip)"),
    depAirportCd: z
      .string()
      .length(3)
      .describe("Departure airport code (3-letter, e.g. HND)"),
    arrAirportCd: z
      .string()
      .length(3)
      .describe("Arrival airport code (3-letter, e.g. SPK)"),
    depDate: z
      .string()
      .regex(/^\d{4}\/\d{2}\/\d{2}$/, "Date must be yyyy/MM/dd format")
      .describe("Departure date in yyyy/MM/dd format"),
    boardingClassCd: z
      .enum(["N", "S"])
      .default("N")
      .describe("Boarding class: N (Normal) or S (Special, +5000 yen)"),
  })
  .strict()
  .refine((d) => d.depAirportCd !== d.arrAirportCd, {
    message: "Departure and arrival airports must be different",
  });

type SearchFlightsInput = z.infer<typeof SearchFlightsSchema>;

/** Shape returned by ATRS GET /flight endpoint. */
interface FareTypeResource {
  fareTypeName: string;
  fare: number;
  vacantNum: number;
}

interface FlightResource {
  flightName: string;
  depAirportName: string;
  arrAirportName: string;
  depTime: string;
  arrTime: string;
  depDate: string;
  boardingClassCd: string;
  fareTypes: Record<string, FareTypeResource>;
}

export function registerFlightTools(server: McpServer): void {
  server.registerTool(
    "atrs_search_flights",
    {
      title: "Search ATRS Flights",
      description: `Search for available flights in the Airline Ticket Reservation System.

Returns flight schedules with vacancy counts and fares per fare-type.

Args:
  - flightType (OW | RT): One-way or round-trip
  - depAirportCd (string): 3-letter departure airport code (e.g. HND, NRT, SPK)
  - arrAirportCd (string): 3-letter arrival airport code
  - depDate (string): Departure date in yyyy/MM/dd format (within 90 days)
  - boardingClassCd (N | S): Normal or Special class (default: N)

Returns:
  List of flights with per-fare-type pricing and seat availability.

Note: Departure and arrival airports must be different. Dates beyond 90 days from today are not searchable.`,
      inputSchema: SearchFlightsSchema,
      annotations: {
        readOnlyHint: true,
        destructiveHint: false,
        idempotentHint: true,
        openWorldHint: false,
      },
    },
    async (params: SearchFlightsInput) => {
      try {
        // Validate airport codes against known set
        if (!AIRPORT_CODES.has(params.depAirportCd)) {
          return {
            content: [
              {
                type: "text" as const,
                text: `Error: Unknown departure airport '${params.depAirportCd}'. Use atrs_list_airports to see valid codes.`,
              },
            ],
          };
        }
        if (!AIRPORT_CODES.has(params.arrAirportCd)) {
          return {
            content: [
              {
                type: "text" as const,
                text: `Error: Unknown arrival airport '${params.arrAirportCd}'. Use atrs_list_airports to see valid codes.`,
              },
            ],
          };
        }

        const flights = await makeApiRequest<FlightResource[]>("flight", "GET", undefined, {
          flightType: params.flightType,
          depAirportCd: params.depAirportCd,
          arrAirportCd: params.arrAirportCd,
          depDate: params.depDate,
          boardingClassCd: params.boardingClassCd,
        });

        if (!flights || flights.length === 0) {
          return {
            content: [
              {
                type: "text" as const,
                text: `No flights found from ${params.depAirportCd} to ${params.arrAirportCd} on ${params.depDate} (class: ${params.boardingClassCd}).`,
              },
            ],
          };
        }

        // Build markdown table
        const lines: string[] = [
          `# Flight Search Results`,
          `**${params.depAirportCd} -> ${params.arrAirportCd}** | ${params.depDate} | Class: ${params.boardingClassCd === "N" ? "Normal" : "Special"}`,
          "",
        ];

        for (const f of flights) {
          lines.push(`## ${f.flightName}  (${f.depTime} - ${f.arrTime})`);
          lines.push("");
          lines.push("| Fare Type | Price (JPY) | Seats Available |");
          lines.push("|-----------|------------|-----------------|");
          for (const [code, ft] of Object.entries(f.fareTypes)) {
            lines.push(
              `| ${ft.fareTypeName} (${code}) | ${ft.fare.toLocaleString()} | ${ft.vacantNum} |`
            );
          }
          lines.push("");
        }

        let text = lines.join("\n");
        if (text.length > CHARACTER_LIMIT) {
          text =
            text.slice(0, CHARACTER_LIMIT) +
            "\n\n... (truncated — narrow search to see all results)";
        }

        return { content: [{ type: "text" as const, text }] };
      } catch (error) {
        return {
          content: [{ type: "text" as const, text: handleApiError(error) }],
        };
      }
    }
  );
}
